### PR TITLE
`Programming exercises`: Improve feedback view

### PR DESCRIPTION
**Pull request by simon-hng**
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->
<!-- # Stacked PR of https://github.com/ls1intum/Artemis/pull/5898 -->
### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Improves feedback view so that students can better understand how their final grade for an exercise is calculated

### Description
<!-- Describe your changes in detail -->
Further refactoring by splitting responsibilities and adding classes, interfaces etc.

#### Introduces a "degenerated" composite pattern for the grouping
We use a composite pattern here in a degenerated form to have a component (`FeedbackNode.component.ts`) 
that can show them independent of whether they are grouped or not.

We can therefore render `FeedbackNodes` for the following types
- [FeedbackGroup, FeedbackGroup, FeedbackGroup]
- [FeedbackItem, FeedbackItem, FeedbackItem]
- [FeedbackGroup, FeedbackGroup, FeedbackItem], etc

This is necessary, as exercises that are not programming exercises do not support groups yet.
Furthermore, the degeneration of this pattern does not allow further nesting of groups.



![image](https://user-images.githubusercontent.com/25489463/206216578-ec3d13cb-f800-4295-840d-015d063b88b8.png)

Objects implementing `FeedbackItemNode` can then be viewed through the `FeedbackItemNodeComponent`

#### Introduces abstractions for methods on `FeedbackItemNodes`
Abstracts creation and grouping of `FeedbackItems` to support future extensions for different exercise types, feedback item types and feedback item groupings

![image](https://user-images.githubusercontent.com/25489463/206218643-01a85758-4bb1-4cd7-8857-9f8f0050fb8f.png)

#### Design changes
Changes the feedback view to resemble the proposed redesign in [figma](https://www.figma.com/proto/nvGhFSvq76iM0TEJCauew6/Artemis?node-id=2615%3A887&starting-point-node-id=2615%3A887)

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

Prerequisites:
- Graded exercise of each type

1. Log in to Artemis
2. Navigate to a graded exercise
3. Test the new feedback view

<!-- #### Exam Mode Testing -->
<!-- If this PR changes some components that are also used in the exam mode, the PR needs additional testing that the exam mode is still working as expected. -->
<!-- If the testing steps above already describe the exam mode or the exam mode cannot be affected by this PR in any way, you can leave this out. -->

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- Note: You may use the table below or copy the file coverage from the Codecov bot's comment. -->

| Class/File | Branch | Line |
|------------|-------:|-----:|
| feedback-service.ts                    |    85.71 |     100 |                                                                                                                                                                                                                                                                                                                     
| feedback.component.ts                  |    83.33 |   95.09 |                                                                                                                                                                                                                                                                                                                     
| feedback-chart.service.ts              |    89.18 |     100 |                                                                                                                                                                                                                                                                                                                     
| feedback-collapse.component.ts         |      100 |     100 |                                                                                                                                                                                                                                                                                                                     
| feedback-group.ts                      |       75 |     100 |                                                                                                                                                                                                                                                                                                                     
| feedback-groups.ts                     |      100 |     100 |                                                                                                                                                                                                                                                                                                                     
| programming-feedback-groups.ts         |      100 |     100 |                                                                                                                                                                                                                                                                                                                     
| feedback-item-service.ts               |       75 |     100 |                                                                                                                                                                                                                                                                                                                     
| programming-feedback-item.service.ts   |    73.17 |     100 |                                                                                                                                                                                                                                                                                                                     
| feedback-node.component.ts             |      100 |     100 |


### Screenshots
![Screenshot_20221219_163758](https://user-images.githubusercontent.com/25489463/208463244-92a65b0e-7013-4e4c-8b31-029d253ab2e7.png)

### CODE REVIEW

**Review by timo07k — COMMENTED** [review]
Everything is working, but I think it is a bit confusing that test cases like you can see below are stated under "Missing":

image

Tested on TS1.

File: src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { Color, ScaleType } from '@swimlane/ngx-charts';
+import { NgxChartsMultiSeriesDataEntry } from 'app/shared/chart/ngx-charts-datatypes';
+import { FeedbackNode } from 'app/exercises/shared/feedback/node/feedback-node';
+import { Exercise } from 'app/entities/exercise.model';
+import { round } from 'app/shared/util/utils';
+import { Injectable } from '@angular/core';
+import { ChartData } from 'app/exercises/shared/feedback/chart/feedback-chart-data';
+
+@Injectable({ providedIn: 'root' })
+export class FeedbackChartService {
```
Comment (ge65cer) [inline src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts:10]: Optional: The contrast for the chart legend and the collapsible title is quite low. This could be enhanced.

Screenshot 2022-12-09 at 10 59 45

**Review by ge65cer — COMMENTED** [review]
It seems the client tests failed.

File: src/main/webapp/app/exercises/shared/feedback/collapse/feedback-collapse.component.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    selector: 'jhi-feedback-collapse',
+    styleUrls: ['./feedback-collapse.scss'],
+    templateUrl: './feedback-collapse.component.html',
+})
+/**
+ * smallCharacterLimit can be adjusted make smaller or bigger items collapsable
+ * isCollapsed tracks whether an item is currently open or closed
+ * text is any string passed to the component
+ */
+export class FeedbackCollapseComponent implements OnInit {
```
[the remark this answers is not in the document]
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/collapse/feedback-collapse.component.ts:14]: Thats a leftover from the earlier implementation. Also, it isn't a real accordion but rather more like a truncated text https://css-tricks.com/multiline-truncated-text-with-show-more-button/.

I also noticed the accordion afterwards as a possibility for the display of groups in `feedback-node.component.ts` but decided against it since it would require me to change a lot about those components to fit the mock

[the remark this answers is not in the document]
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/collapse/feedback-collapse.component.ts:14]: Tests fail due to missing coverage. I need to think of ways to write meaningful tests

File: src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts (hunk 1 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts:10]: Thanks noticed that too but didn't get to it yet

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.html [hunk 3]
```diff
@@ -42,52 +42,53 @@ <h4>
                         </span>
                     </ng-template>
                 </h4>
+                <!--TODO Add badge-->
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/feedback/feedback.component.html:45]: Are you aware of this todo?

File: src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        const absCredits = Math.abs(credits);
+        const absMaxCredits = Math.abs(maxCredits);
+        return Math.sign(credits) * Math.min(absCredits, absMaxCredits);
+    };
+
+    private calculatePercentage = (node: FeedbackNode, maxPoints: number) => {
+        const appliedCredits = this.capCredits(node.credits ?? 0, node.maxCredits);
+        return this.roundToDecimals((appliedCredits / maxPoints) * 100, 2);
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts:129]: We have a rounding function in utils.ts to adhere to the specified accuracy in the course

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    readonly BuildLogType = BuildLogType;
+    readonly AssessmentType = AssessmentType;
+    readonly ExerciseType = ExerciseType;
+    readonly resultIsPreliminary = resultIsPreliminary;
+    readonly roundValueSpecifiedByCourseSettings = roundValueSpecifiedByCourseSettings;
+    readonly xAxisFormatting = axisTickFormattingWithPercentageSign;
+
+    @Input() exercise?: Exercise;
+    @Input() result: Result;
+    // Specify the feedback.text values that should be shown, all other values will not be visible.
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:51]: Why did you not use /** ... */ here?

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+})
+export class FeedbackComponent implements OnInit {
+    readonly BuildLogType = BuildLogType;
+    readonly AssessmentType = AssessmentType;
+    readonly ExerciseType = ExerciseType;
+    readonly resultIsPreliminary = resultIsPreliminary;
+    readonly roundValueSpecifiedByCourseSettings = roundValueSpecifiedByCourseSettings;
+    readonly xAxisFormatting = axisTickFormattingWithPercentageSign;
+
+    @Input() exercise?: Exercise;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:49]: Is the exercise actually optional here?

File: src/main/webapp/app/exercises/shared/feedback/feedback.util.ts [hunk 7]
```diff
@@ -1,20 +1,61 @@
...
+ */
+export const checkSubsequentFeedbackInAssessment = (feedbacks: Feedback[]) => {
+    const gradingInstructions = {}; // { instructionId: number of encounters }
+    for (const feedback of feedbacks) {
+        if (feedback.gradingInstruction && feedback.gradingInstruction.credits !== 0) {
+            if (gradingInstructions[feedback.gradingInstruction!.id!]) {
+                // this grading instruction is counted before
+                const maxCount = feedback.gradingInstruction.usageCount;
+                const encounters = gradingInstructions[feedback.gradingInstruction!.id!];
+                if (maxCount && maxCount > 0) {
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/feedback/feedback.util.ts:48]:
```suggestion
                if (maxCount) {
```

File: src/main/webapp/app/exercises/shared/feedback/group/programming-feedback-groups.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+class FeedbackGroupWrong extends FeedbackGroup {
+    constructor() {
+        super();
+        this.name = 'wrong';
+        this.color = 'danger';
+        this.open = true;
+    }
+
+    shouldContain(feedbackItem: FeedbackItem): boolean {
+        const isReviewerFeedback = feedbackItem.type === 'Reviewer' && feedbackItem.credits && feedbackItem.credits < 0;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/feedback/group/programming-feedback-groups.ts:25]: Why not use the proper Enums here?

File: src/main/webapp/app/exercises/shared/feedback/item/programming-feedback-item.service.ts [hunk 9]
```diff
@@ -96,32 +66,28 @@ export class ProgrammingExerciseFeedbackItemService implements FeedbackItemServi
...
+    private createScaFeedbackItem(feedback: Feedback, showTestDetails: boolean): FeedbackItem {
         const scaCategory = feedback.text!.substring(STATIC_CODE_ANALYSIS_FEEDBACK_IDENTIFIER.length);
         const scaIssue = StaticCodeAnalysisIssue.fromFeedback(feedback);
         const text = showTestDetails ? `${scaIssue.rule}: ${scaIssue.message}` : scaIssue.message;
-        const previewText = computeFeedbackPreviewText(text);
 
         return {
-            type: FeedbackItemType.Issue,
-            category: this.translateService.instant('artemisApp.result.detail.codeIssue.name'),
+            type: 'Static Code Analysis',
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/feedback/item/programming-feedback-item.service.ts:75]: Why the change to a string instead of an enum?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Left some comments

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.html (hunk 3 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/feedback.component.html:45]: yes

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts (hunk 5 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:51]: Didnt't write that

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts (hunk 6 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:49]: Didnt't write that

File: src/main/webapp/app/exercises/shared/feedback/group/programming-feedback-groups.ts (hunk 8 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/group/programming-feedback-groups.ts:25]: Typescript enums have strange behaviours. i.e. inspecting in dev tools leads to `feedbackItem.type: 0` because enums after compilation are just numbers.
https://www.youtube.com/watch?v=0fTdCSH_QEU

File: src/main/webapp/app/exercises/shared/feedback/item/programming-feedback-item.service.ts (hunk 9 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/item/programming-feedback-item.service.ts:75]: It's a string literal union type, functionally allowing us the same type safety as with an enum but with less typing and the aforementioned advantage in dev tools

File: src/main/webapp/app/exercises/shared/feedback/feedback.util.ts (hunk 7 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/feedback.util.ts:48]: Didn't write that

File: src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts (hunk 4 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts:129]: Will look into that thanks!

File: src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts (hunk 1 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts:10]: Follow up fix, as this PR is already too large

**Comment by simon-hng** [comment]
I did a big oopsie and this PR auto-closed

File: src/main/webapp/app/exercises/programming/shared/instructions-render/service/programming-exercise-instruction.service.ts [hunk 10]
```diff
@@ -66,4 +47,29 @@ export class ProgrammingExerciseInstructionService {
             return { testCaseState: TestCaseState.NO_RESULT, detailed: { successfulTests: [], failedTests: [], notExecutedTests: tests } };
         }
     };
+
+    private separateTests(tests: string[], latestResult: Result) {
+        return tests.reduce(
+            (acc, testName) => {
+                const feedback = latestResult ? latestResult.feedbacks?.find(({ text }) => text === testName) : undefined;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/shared/instructions-render/service/programming-exercise-instruction.service.ts:54]:
```suggestion
                const feedback = latestResult?.feedbacks?.find(({ text }) => text === testName);
```

File: src/main/webapp/app/exercises/programming/shared/utils/programming-exercise.utils.ts [hunk 11]
```diff
@@ -13,6 +13,9 @@ import { AssessmentType } from 'app/entities/assessment-type.model';
 
 const BAMBOO_RESULT_LEGACY_TIMESTAMP = 1557526348000;
 
+/*
+ * This is a legacy check, results before the 24th May are considered legacy.
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/shared/utils/programming-exercise.utils.ts:17]: Of which year?

File: src/main/webapp/app/exercises/shared/feedback/node/feedback-node.component.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    feedbackItemGroup: FeedbackGroup;
+
+    // Icons
+    faExclamationTriangle = faExclamationTriangle;
+
+    ngOnInit(): void {
+        if (isFeedbackGroup(this.feedbackItemNode)) {
+            this.feedbackItemGroup = this.feedbackItemNode;
+        } else {
+            this.feedbackItem = <FeedbackItem>this.feedbackItemNode;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/feedback/node/feedback-node.component.ts:31]:
```suggestion
            this.feedbackItem = this.feedbackItemNode as FeedbackItem;
```

File: src/main/webapp/app/exercises/programming/shared/utils/programming-exercise.utils.ts (hunk 11 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/programming/shared/utils/programming-exercise.utils.ts:17]: Its in a comment in the beginning of this class

File: src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+import { Exercise } from 'app/entities/exercise.model';
+import { roundScorePercentSpecifiedByCourseSettings } from 'app/shared/util/utils';
+import { Injectable } from '@angular/core';
+import { ChartData } from 'app/exercises/shared/feedback/chart/feedback-chart-data';
+
+@Injectable({ providedIn: 'root' })
+export class FeedbackChartService {
+    create = (feedbackNodes: FeedbackNode[], exercise: Exercise): ChartData => {
+        const maxPoints = exercise.maxPoints! + (exercise.bonusPoints ?? 0);
+        const xScaleMax = Math.max(100, maxPoints);
```
[the remark this answers is not in the document]
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts]: removed

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.html [hunk 14]
```diff
@@ -23,71 +23,77 @@ <h6 [innerHTML]="'artemisApp.result.afterDueDateFeedbackHidden' | artemisTransla
...
-                                        : {
-                                              score: roundValueSpecifiedByCourseSettings(result.score, course),
-                                              points: roundValueSpecifiedByCourseSettings((result.score * exercise.maxPoints) / 100, course)
-                                          }
-                            }}
+        <div *ngIf="!loadingFailed && feedbackItemNodes && feedbackItemNodes.length; else buildLogsContainer" class="result-detail-container">
+            <div>
+                <div class="d-flex justify-content-between align-items-start">
+                    <h4>
+                        <span *ngIf="!exercise?.maxPoints; else scoreWithPoints">
```
[the remark this answers is not in the document]
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/feedback.component.html:30]: It's not possible to create exercises with 0 points so this case should be covered.

image

**Review by Strohgelaender — APPROVED** [review]
code

**Review by ge65cer — DISMISSED** [review]
Code approved.

**Review by TheZoker — APPROVED** [review]
Tested on TS1 and looks good!

File: src/main/webapp/app/exercises/shared/result/result.service.ts [hunk 15]
```diff
@@ -24,6 +24,12 @@ export type EntityResponseType = HttpResponse<Result>;
 export type EntityArrayResponseType = HttpResponse<Result[]>;
 export type ResultsWithPointsArrayResponseType = HttpResponse<ResultWithPointsPerGradingCriterion[]>;
 
+export interface Badge {
+    class: string;
```
Comment (1999Markus) [inline src/main/webapp/app/exercises/shared/result/result.service.ts:28]: Could you use another name? It clashes with "class" from TS

File: src/test/javascript/spec/component/shared/feedback/feedback-collapse.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
+import { FeedbackCollapseComponent } from 'app/exercises/shared/feedback/collapse/feedback-collapse.component';
+
+describe('FeedbackCollapseComponent', () => {
+    /*
+     * Same value as in feedback-collapse.component.ts
+     */
+    const FEEDBACK_PREVIEW_CHARACTER_LIMIT = 300;
```
Comment (1999Markus) [inline src/test/javascript/spec/component/shared/feedback/feedback-collapse.component.spec.ts:7]: This value should be at some central place. Duplicating it across the codebase doesn't seem like a good idea

**Review by 1999Markus — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/exercises/shared/result/result.service.ts (hunk 15 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/result/result.service.ts:28]: was that way before and I would prefer not to make any more changes

File: src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+     * Sets credits to 0 for all feedback nodes
+     */
+    private clearCredits = (feedbackNodes: FeedbackNode[]) => {
+        return feedbackNodes.map((node) => ({
+            ...node,
+            credits: 0,
+        }));
+    };
+
+    private capCredits = (credits: number, maxCredits?: number): number => {
```
[the remark this answers is not in the document]
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/feedback/chart/feedback-chart.service.ts:109]: follow up

**Review by 1999Markus — DISMISSED** [review]

**Comment by 1999Markus** [comment]
code

**Review by stefanwaldhauser — DISMISSED** [review]
Approved! Testing in testing session :)

**Review by akesfeden — DISMISSED** [review]
Tested on TS1 in testing session, new feedback view looks good and I didn't notice any errors.

**Review by tunargs — DISMISSED** [review]
Tested during the testing session (02.01.2022), looks perfect!

**Review by pellumbbaboci — APPROVED** [review]
Tested during testing session on ts1, works as expected.

**Review by Strohgelaender — APPROVED** [review]

**Review by TheZoker — APPROVED** [review]

**Review by stefanwaldhauser — APPROVED** [review]
Reapprove

**Review by ge65cer — APPROVED** [review]
